pkg/clusteragent/admission/mutate/common.go

```diff
@@ -153,3 +159,67 @@ func containsVolumeMount(volumeMounts []corev1.VolumeMount, element corev1.Volum
 	}
 	return false
 }
+
+// shouldInject returns true if Admission Controller should inject standard tags, APM configs and APM libraries
+func shouldInject(pod *corev1.Pod) bool {
+	// If a pod explicitly sets the label admission.datadoghq.com/enabled, make a decision based on its value
+	if val, found := pod.GetLabels()[admCommon.EnabledLabelKey]; found {
+		switch val {
+		case "true":
+			return true
+		case "false":
+			return false
+		default:
+			log.Warnf("Invalid label value '%s=%s' on pod %s should be either 'true' or 'false', ignoring it", admCommon.EnabledLabelKey, val, podString(pod))
+		}
+	}
+
+	return isApmInstrumentationEnabled(pod.GetNamespace()) || config.Datadog.GetBool("admission_controller.mutate_unlabelled")
+}
+
+// isApmInstrumentationEnabled indicates if Single Step Instrumentation is enabled for the namespace in the cluster
+// Single Step Instrumentation is enabled if:
+//   - cluster-wide configuration `apm_config.instrumentation.enabled` is true and the namespace is not excluded via `apm_config.instrumentation.disabled_namespaces`
+//   - cluster-wide configuration `apm_config.instrumentation.enabled` is false and the namespace is included via `apm_config.instrumentation.enabled_namespaces`
+func isApmInstrumentationEnabled(namespace string) bool {
+	apmInstrumentationEnabled := config.Datadog.GetBool("apm_config.instrumentation.enabled")
+	isNsEnabled := apmInstrumentationEnabled
+
+	apmEnabledNamespaces := config.Datadog.GetStringSlice("apm_config.instrumentation.enabled_namespaces")
+	apmDisabledNamespaces := config.Datadog.GetStringSlice("apm_config.instrumentation.disabled_namespaces")
+
+	// If Single Step Instrumentation is enabled in the cluster, enabling it in the specific namespaces is redundant
+	if apmInstrumentationEnabled && len(apmEnabledNamespaces) > 0 {
+		log.Warnf("Redundant configuration option `apm_config.instrumentation.enabled_namespaces:%v`", apmEnabledNamespaces)
+	}
+
+	// If Single Step Instrumentation is disabled in the cluster, disabling it in the specific namespaces is redundant
+	if !apmInstrumentationEnabled && len(apmDisabledNamespaces) > 0 {
+		log.Warnf("Redundant configuration option `apm_config.instrumentation.disabled_namespaces:%v`", apmDisabledNamespaces)
+	}
+
+	// apm.instrumentation.enabled_namespaces and apm.instrumentation.disabled_namespaces configuration cannot be set at the same time
+	if len(apmEnabledNamespaces) > 0 && len(apmDisabledNamespaces) > 0 {
+		log.Errorf("apm.instrumentation.enabled_namespaces and apm.instrumentation.disabled_namespaces configuration cannot be set together")
+		return false
+	}
+
+	// Single Step Instrumentation for the namespace can override cluster-wide configuration
+	for _, ns := range apmEnabledNamespaces {
+		if ns == namespace {
+			return true
+		}
+	}
+
+	// Unless kube-system ns is explicitly enabled, always disable Single Step Instrumentation
+	if namespace == namespaceWithAlwaysDisabledInjections {
```

Comment:
I suspect that at some point in the future, there will be more than one namespace in this category and this logic will have to be tweaked. But I am not a fan of speculative coding, so this is fine for now.

Reply:
For the next pass on this logic I'm going to follow @vboulineau 's suggestion https://github.com/DataDog/datadog-agent/pull/19504#discussion_r1338842124 and implement more complicated filtering logic for enabled/disabled namespaces. For the sake of getting SSI out before the code freeze, I'm going to merge the code as is.